lib/linked_list.rb

```diff
@@ -2,44 +2,81 @@
 
 class LinkedList
   attr_reader :head
-
+  
   def initialize
     @head = nil
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def add_first(data)
-
+    @head = Node.new(data, @head)
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    return nil if @head.nil?
+    
+    return @head.data
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def length
-    return 0
+    current = @head
+    length = 0
+    
+    while current
+      length += 1
+      current = current.next
+    end
+    
+    return length
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def add_last(data)
-
+    if @head.nil?
+      return @head = Node.new(data)
+    end
+    
+    previous = nil
+    node = @head
+    
+    while node
+      previous = node
+      node = node.next
+    end
+    previous.next = Node.new(data)
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def get_last
-
+    previous = nil
```

Comment:
What if `head` is `nil`?

```suggestion
    return nil if head.nil?
    previous = nil
```